File: packages/aegis-sdk-py/salanor_aegis_ledger/schema.py
from __future__ import annotations

import json
from pathlib import Path

from jsonschema import Draft202012Validator

_REPO_ROOT = Path(__file__).resolve().parents[3]
_SCHEMA_PATH = _REPO_ROOT / "spec" / "aps" / "v0.1.json"


class ApsValidationError(ValueError):
    def __init__(self, message: str, details: list[str] | None = None):
        super().__init__(message)
        self.details = details or []


def _validator() -> Draft202012Validator:
    schema = json.loads(_SCHEMA_PATH.read_text(encoding="utf-8"))
    return Draft202012Validator(schema)


def validate_event(event: object) -> None:
    validator = _validator()
    errors = sorted(validator.iter_errors(event), key=lambda e: e.path)
    if errors:
        details = [f"/{'/'.join(str(p) for p in err.path)} {err.message}" for err in errors]
        raise ApsValidationError("APS event failed schema validation", details)

File: packages/aegis-ledger-sdk/src/schema.ts
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import Ajv2020 from "ajv/dist/2020";
import addFormats from "ajv-formats";
import type { ApsEvent } from "./types.js";

const schemaDir = join(dirname(fileURLToPath(import.meta.url)), "../../../spec/aps");
const apsSchema = JSON.parse(readFileSync(join(schemaDir, "v0.1.json"), "utf8"));

const ajv = new Ajv2020({ allErrors: true, strict: false });
addFormats(ajv);
const validateAps = ajv.compile(apsSchema);

export class ApsValidationError extends Error {
  constructor(
    message: string,
    readonly details: string[] = [],
  ) {
    super(message);
    this.name = "ApsValidationError";
  }
}

export function validateEvent(event: unknown): ApsEvent {
  if (!validateAps(event)) {
    const details =
      validateAps.errors?.map((e) => `${e.instancePath || "/"} ${e.message ?? "invalid"}`) ?? [];
    throw new ApsValidationError("APS event failed schema validation", details);
  }
  return event as ApsEvent;
}
